Core/include/Acts/Clusterization/InPlaceClusterization.hpp

```diff
@@ -0,0 +1,371 @@
+template <typename coordinates_t>
+bool isConnectedCommonEdgeOrCorner(const coordinates_t &coordinates_diff) {
+  bool connected = true;
+  for (const auto &a_coordinate_diff : coordinates_diff) {
+    connected &= a_coordinate_diff <= 1;
+  }
+  return connected;
+}
+
+/// @brief test whether cells are connected considering common edges only.
+/// @param coordinates_diff absolute values of the differences of the cell coordinates
+/// Test whether two cells with the given coordinate differences are connected
+/// either horizontally, vertically or are equal.
+template <typename coordinates_t>
+bool isConnectedCommonEdge(const coordinates_t &coordinates_diff) {
+  int connections = 0;
+  for (const auto &a_coordinate_diff : coordinates_diff) {
+    connections += a_coordinate_diff;
+  }
+  return connections <= 1;
+}
+
+/// @brief test whether cells are connected considering common edges or common corners
+enum class EConnectionType { CommonEdgeOrCorner, CommonEdge };
+template <EConnectionType CONNECTION_TYPE, typename coordinates_t>
+bool isConnected(const coordinates_t &coordinates_diff) {
+  if constexpr (CONNECTION_TYPE == EConnectionType::CommonEdgeOrCorner) {
+    return isConnectedCommonEdgeOrCorner(coordinates_diff);
+  } else {
+    return isConnectedCommonEdge(coordinates_diff);
+  }
+}
+
+/// @brief test whether the search for connections can be aborted.
+/// the search is performed in descending order in one coordinate, once the
+/// difference in this coordinate is too large the search can be aborted because
+/// there cannot be any further connections.
+template <typename coordinates_t>
+bool canAbortSearch(const coordinates_t &coordinates_diff,
+                    unsigned int sort_axis_i) {
+  assert(sort_axis_i < coordinates_diff.size());
+  return coordinates_diff[sort_axis_i] > 1;
+}
+
+/// @brief compute the absolute difference of two coordinates
+template <typename coordinate_t>
+auto absDifference(coordinate_t a, coordinate_t b) {
+  if constexpr (std::is_signed_v<coordinate_t>) {
+    return std::abs(a - b);
+  } else {
+    return (a > b ? a - b : b - a);
+  }
+}
+
+/// @brief Label cells which are in ascending order of the coordinate of the given axis
+/// @param cells the sorted cell collection that should be labeled.
+template <unsigned int SORT_AXIS, CellCollection cell_collection_t,
+          std::unsigned_integral index_t = unsigned int,
+          EConnectionType CONNECTION_TYPE>
+void labelSortedCells(cell_collection_t &cells) {
+  using cell_t = std::remove_cvref_t<decltype(cells[0])>;
+  static constexpr std::size_t NDim = traits::getCellDimension<cell_t>();
+  static_assert(NDim > 0);
+  using label_t = std::remove_cvref_t<decltype(traits::getLabel(cells[0]))>;
+  using coordinate_t =
+      std::remove_cvref_t<decltype(traits::getCellCoordinate(cells[0], 0u))>;
+  static_assert(std::numeric_limits<index_t>::max() >=
+                std::numeric_limits<label_t>::max());
```

Comment:
also these look like they can be ensured / provided by the `CellCollection` concept